### PR TITLE
[AMORO-508]: FlinkOptimizerContainer support flink native kubernetes applicatoin

### LINKED ISSUE #508: [Feature]: Support Flink Optimizer on kubernetes
### Description

`FlinkOptimizer` is designed to run on YARN. We need to support running the Flink optimizer on kubbernetes.

### Use case/motivation

_No response_

### Describe the solution

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by baiyangtx**
## Why are the changes needed?

Close #508.

## Brief change log

- Make `FlinkOptimizerContainer` support flink native kubernetes application. 
- Introducing multi-level configuration for flink-conf, allowing Flink configurations to take effect according to the priority of group -> container -> flink-conf.yaml. 
- Support configuring `ams-optimizing-uri` in Container or Group to enable production scenarios that cannot be met by `server-expose-host`. 
- Add Dockerfile for optimizer-flink image

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [X] Add screenshots for manual tests if appropriate

- [ ] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes  )
- If yes, how is the feature documented? (docs)

### CODE REVIEW

File: ams/api/src/main/java/com/netease/arctic/ams/api/Environments.java [hunk 1]
```diff
@@ -15,7 +15,11 @@ public static String getHomePath() {
     if (arcticHome != null) {
       return arcticHome;
     }
-    return System.getProperty(AMORO_HOME);
+    arcticHome = System.getProperty(AMORO_HOME);
```
Comment (huyuanfeng2018) [inline ams/api/src/main/java/com/netease/arctic/ams/api/Environments.java:18]: should be amoroHome

Comment (baiyangtx) [reply ams/api/src/main/java/com/netease/arctic/ams/api/Environments.java]: Got it！

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java [hunk 2]
```diff
@@ -35,52 +41,90 @@
...
+    this.target = Target.valueToEnum(runTarget);
+    String jobUri = containerProperties.get(FLINK_JOB_URI);
+    if (target.applicationMode) {
+      Preconditions.checkArgument(
+          StringUtils.isNotEmpty(jobUri),
+          "FlinkOptimizerContainer init error, the property " +
+              FLINK_JOB_URI + " is required if run target in " + "application mode.");
+    }
+    if (StringUtils.isEmpty(jobUri)) {
+      jobUri = amsHome + "/plugin/optimize/OptimizeJob.jar";
```
Comment (zhoujinsong) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java:107]: It is recommended to use a constant variables to store the default value.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java [hunk 3]
```diff
@@ -35,52 +41,90 @@
...
+      ).build();
 
-  public String getFlinkHome() {
-    return Optional.ofNullable(PropertyUtil.checkAndGetProperty(getContainerProperties(), FLINK_HOME_PROPERTY))
-        .map(flinkHome -> flinkHome.replaceAll("/$", ""))
-        .orElse(null);
+      String imageRef = flinkConf.configValue(FlinkConf.PropertyKeys.KUBERNETES_IMAGE_REF);
+      Preconditions.checkArgument(
+          StringUtils.isNotEmpty(imageRef),
+          "FlinkOptimizerContainer init error, The flink-conf: " +
```
Comment (zhoujinsong) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java:120]: Pay attention to ge grammer issues in log message.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/util/FlinkConf.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  public static class PropertyKeys {
+    public static final String JOB_MANAGER_TOTAL_PROCESS_MEMORY = "jobmanager.memory.process.size";
+    public static final String TASK_MANAGER_TOTAL_PROCESS_MEMORY = "taskmanager.memory.process.size";
+    public static final String KUBERNETES_IMAGE_REF = "kubernetes.container.image";
+
+    public static final String KUBERNETES_CLUSTER_ID = "kubernetes.cluster-id";
+    public static final String KUBERNETES_TASKMANAGER_LABLES = "kubernetes.taskmanager.labels";
+    public static final String KUBERNETES_JOBMANAGER_LABLES = "kubernetes.jobmanager.labels";
+
+    public static final String KUBERNETES_NAMESPACE = "kubernetes.namespace";
```
Comment (zhoujinsong) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/util/FlinkConf.java:22]: Seems to be not used.

File: docs/admin-guides/managing-optimizers.md [hunk 5]
```diff
@@ -32,32 +32,62 @@ containers:
...
+| target        | true     | yarn-pre-job | flink job deployed target, available values `yarn-pre-job`, `yarn-application`, `kubernetes-application`                                                                 |
+| job-uri       | false    | N/A      | The jar uri of flink optimizer job. This is required if target is application mode.                                                                                      |
+| ams-optimizing-uri| false | N/A       | uri of AMS thrift self-optimizing endpoint. This could be used if the ams.server-expose-host is not available                                                            |
+| export.\<key\> | false | N/A     | environment variables will be exported during job submit                                                                                                                 |
+| flink-conf.\<key\> | false | N/A     | [Flink Configuration Options](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/config/) will be passed to cli by `-Dkey=value`, |
+
+Environment variables:
```
Comment (zhoujinsong) [inline docs/admin-guides/managing-optimizers.md:49]: Environment variables are also part of properties. Should we introduce them in the same table?

File: docs/admin-guides/managing-optimizers.md [hunk 6]
```diff
@@ -84,11 +114,19 @@ The following configuration needs to be filled in:
...
+| memory   | local | Yes      | N/A                   | The memory size of the local optimizer Java process.                                                                                                                                                                                                                                                                                                                                                             |
+| jobmanager.memory | flink | No       | N/A                   | The memory size of flink jobmanger, it will overwrite the `jobmanager.memory.process.size`.                                                                                                                                                                                                                                                                                                                      |
```
Comment (zhoujinsong) [inline docs/admin-guides/managing-optimizers.md:121]: We have deprecated these 2 properties.

File: docs/admin-guides/managing-optimizers.md [hunk 7]
```diff
@@ -84,11 +114,19 @@ The following configuration needs to be filled in:
...
+| taskmanager.memory | flink | No | N/A                   | The memory size of flink taskmanger, it will overwrite the `taskmanger.memory.process.size`.                                                                                                                                                                                                                                                                                                                     |
+
+Tips, the following properties of OptimizerContainer could be overwritten by group properties.
+
+| Properties         | Container Type | Description                                   |
```
Comment (zhoujinsong) [inline docs/admin-guides/managing-optimizers.md:126]: Is it necessary to split it into different tables for display here?

File: ams/api/src/main/java/com/netease/arctic/ams/api/Environments.java [hunk 8]
```diff
@@ -7,15 +7,19 @@ public class Environments {
...
+    amoroHome = System.getenv(AMORO_HOME);
+    if (amoroHome != null) {
+      return amoroHome;
     }
-    return System.getProperty(AMORO_HOME);
+    amoroHome = System.getProperty(AMORO_HOME);
+    if (amoroHome != null) {
+      return amoroHome;
+    }
+    return System.getProperty("user.dir");
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/api/src/main/java/com/netease/arctic/ams/api/Environments.java:22]: Default value of `user.dir` is the location where you call the `java` command. 

I have set to static final var

File: ams/api/src/main/java/com/netease/arctic/ams/api/resource/Resource.java [hunk 9]
```diff
@@ -118,5 +112,12 @@ public Builder setProperties(Map<String, String> properties) {
       this.properties = properties;
       return this;
     }
+
+    // In some cases, shorter strings are needed for UUIDs.
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/api/src/main/java/com/netease/arctic/ams/api/resource/Resource.java:116]: Add addition comments.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java [hunk 10]
```diff
@@ -35,52 +41,90 @@
...
+  public void init(String name, Map<String, String> containerProperties) {
+    super.init(name, containerProperties);
+    this.flinkHome = getFlinkHome();
+    this.flinkConfDir = getFlinkConfDir();
+
+    String runTarget = Optional.ofNullable(containerProperties.get(FLINK_RUN_TARGET))
+        .orElse(Target.YarnPreJob.value);
+    this.target = Target.valueToEnum(runTarget);
+    String jobUri = containerProperties.get(FLINK_JOB_URI);
+    if (target.applicationMode) {
```
[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java:100]: it's a final field and target is an Enum

File: docs/admin-guides/managing-optimizers.md (hunk 5 above)
Comment (baiyangtx) [reply docs/admin-guides/managing-optimizers.md:49]: export.<key>  has been listed in table. 

The following part is used to show some important environment variables.

File: docs/admin-guides/managing-optimizers.md (hunk 7 above)
Comment (baiyangtx) [reply docs/admin-guides/managing-optimizers.md:126]: Here is properties guide for OptimizerGroup, the above table is properties guide for OptimizerContainer.

**Review by zhoujinsong — APPROVED** [review]
LGTM.

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java [hunk 11]
```diff
@@ -216,21 +290,102 @@ private String readApplicationId(Process exec) {
...
+    }
+    return this.flinkHome + FLINK_CONFIG_PATH;
+  }
+
+  private String kubernetesClusterId(Resource resource) {
+    return "amoro-optimizer-" + resource.getResourceId();
+  }
+
+  private enum Target {
+    YarnPreJob("yarn-per-job", false),
```
Comment (zhongqishang) [inline ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java:366]: Upper case and under line?

File: dist/src/main/arctic-bin/conf/config.yaml [hunk 12]
```diff
@@ -95,7 +95,12 @@ containers:
 #    container-impl: com.netease.arctic.optimizer.FlinkOptimizerContainer
 #    properties:
 #      flink-home: "/opt/flink/"                                     # Flink install home
+#      target: "yarn-pre-job"                                        # Flink run target, (yarn-pre-job, yarn-application, kubernetes-application)
```
Comment (zhongqishang) [inline dist/src/main/arctic-bin/conf/config.yaml:98]: -> `yarn-per-job`

File: docs/admin-guides/managing-optimizers.md [hunk 13]
```diff
@@ -32,32 +32,64 @@ containers:
...
+|---------------------------|----------|---------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| flink-home                | true     | N/A           | Flink installation location                                                                                                                                                                                                                                                          |
+| target                    | true     | yarn-pre-job  | flink job deployed target, available values `yarn-pre-job`, `yarn-application`, `kubernetes-application`                                                                                                                                                                             |
```
Comment (zhongqishang) [inline docs/admin-guides/managing-optimizers.md:43]: -> `yarn-per-job`

File: docs/admin-guides/managing-optimizers.md [hunk 14]
```diff
@@ -32,32 +32,64 @@ containers:
...
+
+{{< hint info >}}
+To better utilize the resources of Flink Optimizer, it is recommended to add the following configuration to the Flink Optimizer Group:
+* Set `flink-conf.taskmanager.memory.managed.size` to `32mb` as Flink optimizer does not have any computation logic, it does not need to occupy managed memory.
+* Set `flink-conf.taskmanager.memory.netwrok.max` to `32mb` as there is no need for communication between operators in Flink Optimizer.
+* Set `flink-conf.taskmanager.memory.netwrok.nin` to `32mb` as there is no need for communication between operators in Flink Optimizer.
+{{< /hint >}}
+
+
+An example for yarn-pre-job mode:
```
Comment (zhongqishang) [inline docs/admin-guides/managing-optimizers.md:62]: -> `yarn-per-job`

File: docs/admin-guides/managing-optimizers.md [hunk 15]
```diff
@@ -32,32 +32,64 @@ containers:
...
     properties:
-      flink-home: /opt/flink/        #flink install home
-      export.HADOOP_CONF_DIR: /etc/hadoop/conf/       #hadoop config dir
-      export.HADOOP_USER_NAME: hadoop       #hadoop user submit on yarn
+      flink-home: /opt/flink/                              #flink install home
+      export.HADOOP_CONF_DIR: /etc/hadoop/conf/            #hadoop config dir
+      export.HADOOP_USER_NAME: hadoop                      #hadoop user submit on yarn
       export.JVM_ARGS: -Djava.security.krb5.conf=/opt/krb5.conf       #flink launch jvm args, like kerberos config when ues kerberos
-      export.FLINK_CONF_DIR: /etc/hadoop/conf/        #flink config dir
+      export.FLINK_CONF_DIR: /etc/hadoop/conf/              #flink config dir
```
Comment (zhongqishang) [inline docs/admin-guides/managing-optimizers.md:73]: -> `/opt/flink/conf/`

File: docs/admin-guides/managing-optimizers.md [hunk 16]
```diff
@@ -32,32 +32,64 @@ containers:
...
+```yaml
+containers:
+  - name: flinkContainer
+    container-impl: com.netease.arctic.optimizer.FlinkOptimizerContainer
+    properties:
+      flink-home: /opt/flink/                                                        #flink install home
+      target: kubernetes-application                                                 #flink run as native kubernetes
+      job-uri: "local:///opt/flink/usrlib/OptimizeJob.jar"                           #optimizer job location in image
+      ams-optimizing-uri: thrift://ams.amoro.service.local:1261                      #AMS optimizing uri 
+      export.FLINK_CONF_DIR: /etc/hadoop/conf/                                       #flink config dir
```
Comment (zhongqishang) [inline docs/admin-guides/managing-optimizers.md:87]: -> /opt/flink/conf/

Comment (baiyangtx) [reply docs/admin-guides/managing-optimizers.md]: fixed

File: docs/admin-guides/managing-optimizers.md (hunk 15 above)
Comment (baiyangtx) [reply docs/admin-guides/managing-optimizers.md:73]: fixed

File: docs/admin-guides/managing-optimizers.md (hunk 14 above)
Comment (baiyangtx) [reply docs/admin-guides/managing-optimizers.md]: fixed

File: docs/admin-guides/managing-optimizers.md (hunk 13 above)
Comment (baiyangtx) [reply docs/admin-guides/managing-optimizers.md]: fixed

File: dist/src/main/arctic-bin/conf/config.yaml (hunk 12 above)
Comment (baiyangtx) [reply dist/src/main/arctic-bin/conf/config.yaml]: fixed

File: ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java (hunk 11 above)
Comment (baiyangtx) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java]: fixed

Comment (zhongqishang) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java]: I mean is that enum types are defined using uppercase and underline.
```
    YARN_PER_JOB("yarn-per-job", false),
    YARN_APPLICATION("yarn-application", true),
    KUBERNETES_APPLICATION("kubernetes-application", true);
```

Comment (baiyangtx) [reply ams/optimizer/src/main/java/com/netease/arctic/optimizer/FlinkOptimizerContainer.java]: fixed

**Pull request merged**